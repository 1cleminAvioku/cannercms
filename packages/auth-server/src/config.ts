import { AppConfig } from 'simple-oidc-server/lib/config';
import { Logger } from '@canner/server-common/lib/interface';

export interface AuthConfig extends AppConfig {
  mountPath?: string;
  logger?: Logger;
}
