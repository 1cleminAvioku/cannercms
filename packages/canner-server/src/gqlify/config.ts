
export interface GqlifyConfig {
  schemaPath?: string;
  dataSources?: Record<string, any>;
}
