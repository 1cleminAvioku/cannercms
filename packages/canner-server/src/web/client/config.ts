import path from 'path';

export const HTML_PATH = path.join(__dirname, './statics/index.html');
export const SCHEMA_PATH = path.join(__dirname, './statics/schema/canner.schema.js');
export const SCHEMA_OUTPUT_PATH = process.cwd();
export const WEB_OUTPUT_PATH = path.join(process.cwd(), '.cms');
export const SCHEMA_ONLY = process.env.SCHEMA_ONLY;
export const WEB_ONLY = process.env.WEB_ONLY;
export const CLOUD_PATH = path.join(__dirname, './statics/default.canner.cloud.ts');
