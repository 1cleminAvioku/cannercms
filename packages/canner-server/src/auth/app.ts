import { createApp } from 'simple-oidc-server';
import koaMount from 'koa-mount';
import Koa from 'koa';
import { WebService, Logger } from '../common/interface';
import { AuthConfig } from './config';

export class AuthService implements WebService {
  private logger: Logger;
  private config: AuthConfig;

  constructor(config: AuthConfig) {
    this.config = config;
  }

  public async mount(app: Koa) {
    const mountPath = this.config.mountPath;
    const config: AuthConfig = {
      ...this.config,
      appPrefix: `${mountPath}/`,
    };
    const authApp = await createApp(config);
    app.use(koaMount(mountPath, authApp));
  }
}
