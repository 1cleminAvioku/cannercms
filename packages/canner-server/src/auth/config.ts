import { AppConfig } from 'simple-oidc-server/lib/config';

export {
  AppConfig as AuthConfig
};
