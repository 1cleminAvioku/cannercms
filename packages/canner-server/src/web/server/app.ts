import Koa, {Context} from 'koa';
import path from 'path';
import serve from 'koa-static';
import Router from 'koa-router';
import koaMount from 'koa-mount';
import views from 'koa-views';
import { construct } from './auth/middleware';

// config
import { createConfig, CmsServerConfig } from './config';
import { WebService, Logger } from '../../common/interface';
import { jsonLogger } from '../../common/jsonLogger';

// constants
import { usernameCookieKey, accessTokenCookieKey, authCallbackPath } from './constants';

export class CmsWebService implements WebService {
  private logger: Logger = jsonLogger;
  private config: CmsServerConfig;

  constructor(customConfig?: CmsServerConfig) {
    const config = createConfig(customConfig);
    this.config = config;
  }

  public async mount(rootApp: Koa) {
    const config = this.config;
    const app = new Koa();
    // todo: isolate the cookieKeys under mounted app
    // https://github.com/koajs/mount/pull/58
    app.keys = config.cookieKeys;

    // construct auth related middlewares
    const {
      beforeRenderCms,
      authCallback,
      logout,
    } = await construct(config, authCallbackPath);

    // serve client static
    const serveClientStatic =
      koaMount(this.config.staticsPath, serve(this.config.clientBundledDir, {gzip: true, index: false}));
    app.use(serveClientStatic);

    // serve favicon
    const favicon =
      koaMount('/public/favicon', serve(path.resolve(__dirname, '../public/favicon'), {gzip: true, index: false}));
    app.use(favicon);

    // router
    const router = new Router();
    router.use(views(path.join(__dirname, './views'), {
      extension: 'pug'
    }));

    // cms
    const setConfigMiddleware = async (ctx: Context, next: () => Promise<any>) => {
      // put into frontendConfig
      const username = ctx.cookies.get(usernameCookieKey, {signed: true});
      const accessToken = ctx.cookies.get(accessTokenCookieKey, {signed: true});
      ctx.state.frontendConfig = JSON.stringify({username, accessToken});
      return next();
    };

    router.get('/cms', beforeRenderCms, setConfigMiddleware, async ctx => {
      await ctx.render('cms', {title: 'Canenr CMS', staticsPath: config.staticsPath});
    });

    router.get('/cms/*', beforeRenderCms, setConfigMiddleware, async ctx => {
      await ctx.render('cms', {title: 'Canenr CMS', staticsPath: config.staticsPath});
    });

    // auth callback
    router.get(authCallbackPath, authCallback, async ctx => {
      ctx.redirect('/cms');
    });

    // logout
    router.get('/auth/logout', logout, async ctx => {
      // if logout middleware calls next, it will redirect to /cms
      ctx.redirect('/cms');
    });

    // health check
    router.get('/health', async ctx => {
      ctx.status = 200;
    });

    // redirect
    router.get('/', async (ctx: Context) => {
      return ctx.redirect('/cms');
    });

    app.use(router.routes());

    // mount cmsApp to rootApp
    const cmsMiddleware = koaMount(app);
    rootApp.use(cmsMiddleware);
  }

  // logger
  public setLogger(logger: Logger) {
    this.logger = logger;
  }

  public getLogger(): Logger {
    return this.logger;
  }
}
