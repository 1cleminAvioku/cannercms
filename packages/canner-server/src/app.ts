import Koa from 'koa';
import { construct } from './config/factory';

// Service
import { GraphQLService } from '@canner/graphql-server/lib/app';
import { CmsWebService } from '@canner/cms-server/lib/server/app';
import { AuthService } from '@canner/auth-server/lib/app';
import { ServerConfig } from './config/interface';
import { get } from 'lodash';

export interface RootAppConfig {
  cookieKeys: string[];
}

export const createApp = async (config: ServerConfig) => {
  const app = new Koa();
  const {cmsConfig, graphqlConfig, rootAppConfig, authConfig} = construct(config);

  // todo: isolate cookieKeys for different service
  // https://github.com/koajs/mount/pull/58
  app.keys = rootAppConfig.cookieKeys;

  // construct services
  const cmsWebService = new CmsWebService(cmsConfig);
  await cmsWebService.mount(app);

  // graphql service is optional, it might be created to other server
  if (graphqlConfig) {
    const graphqlService = new GraphQLService(graphqlConfig);
    await graphqlService.mount(app);
  }

  // auth service is optional, might be created to other server as well
  if (authConfig) {
    const authService = new AuthService(authConfig);
    await authService.mount(app);
  }

  return app;
};
