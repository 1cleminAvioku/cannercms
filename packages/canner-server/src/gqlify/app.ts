import path from 'path';

import { ServiceAccount } from 'firebase-admin';
import  { Gqlify, MemoryDataSource } from '@gqlify/server';
import { ScalarField, Model } from '@gqlify/server/lib/dataModel';
import { DataModelType } from '@gqlify/server/lib/dataModel/type';
import { CannerSchemaToGQLifyModel, IDataSourceOption } from '@gqlify/canner-schema-to-gqlify-model';
import  { ApolloServer } from 'apollo-server';
import { readFileSync } from 'fs';

interface ICreateAppOptions {
  schemaPath?: string;
  dataSources?: Record<string, IDataSourceOption>;
}

export const createApp = async (options: ICreateAppOptions = {}): Promise<{app: ApolloServer}> => {
  // Read datamodel
  const schemaPath = path.resolve(process.cwd(), options.schemaPath || 'schema.node.js');
  const cannerSchema = JSON.parse(readFileSync(schemaPath, { encoding: 'utf8' }));
  const cannerSchemaToGQLifyModel = new CannerSchemaToGQLifyModel(cannerSchema, new MemoryDataSource());

  const models: Model[] = Object.values(cannerSchemaToGQLifyModel.models);
  for(const model of models) {
    const idField = model.getField('id');
    if (!idField) {
      model.appendField('id', new ScalarField({ type: DataModelType.ID, nonNull: true, unique: true }))
    }
  }
  // construct gqlify
  const gqlify = new Gqlify({
    rootNode: cannerSchemaToGQLifyModel.rootNode,
    models
  });

  // GQLify will provide GraphQL apis & resolvers to apollo-server
  const server = new ApolloServer(gqlify.createApolloConfig());
  return { app: server };
}
