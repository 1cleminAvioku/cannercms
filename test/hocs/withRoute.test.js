/**
|--------------------------------------------------
| this is just for copy and past
|--------------------------------------------------
*/


import * as React from 'react';
import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import withRoute from '../../src/hocs/withRoute';
import {fromJS} from 'immutable';
import { UNIQUE_ID } from '../../src/app/config';
import RefId from 'canner-ref-id';

Enzyme.configure({ adapter: new Adapter() });

describe('withRoute', () => {
  let WrapperComponent, MockComponent, mockRenderChildren;

  beforeEach(() => {
    MockComponent = function MockComponent() {
      return (<div>Component</div>);
    }
    mockRenderChildren = jest.fn().mockImplementation(() => <div>children</div>);
    WrapperComponent = withRoute(MockComponent);
  });

  it('should render', () => {
    const mockFetch = jest.fn().mockImplementation(() => Promise.resolve());
    const wrapper = shallow(<WrapperComponent
      type="string"
      routes={['info', 'name']}
      rootValue={fromJS({
        name: 'Name'
      })}
      name="name"
      refId={new RefId("info/name")}
      params={{}}
      renderChildren={mockRenderChildren}
      fetch={mockFetch}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(mockFetch).toHaveBeenCalledTimes(1);
  });

  it('normal render when routes = []', () => {
    const mockFetch = jest.fn().mockImplementation(() => Promise.resolve());
    const wrapper = shallow(<WrapperComponent
      type="string"
      routes={[]}
      rootValue={fromJS({
        name: 'Name'
      })}
      name="name"
      refId={new RefId("info/name")}
      params={{}}
      renderChildren={mockRenderChildren}
      fetch={mockFetch}
    />);
    return wrapper.instance().updateState().then(() => {
      expect(wrapper.instance().state).toMatchObject({
        renderType: 0,
        canRender: true,
        restRoutes: []
      });
    });
  });

  it('normal render when routes = [selfName]', () => {
    const mockFetch = jest.fn().mockImplementation(() => Promise.resolve());
    const wrapper = shallow(<WrapperComponent
      type="string"
      routes={['name']}
      rootValue={fromJS({
        name: 'Name'
      })}
      name="name"
      refId={new RefId("info/name")}
      params={{}}
      renderChildren={mockRenderChildren}
      fetch={mockFetch}
    />);
    return wrapper.instance().updateState().then(() => {
      expect(wrapper.instance().state).toMatchObject({
        renderType: 0,
        canRender: true,
        restRoutes: []
      });
    });
  });

  it('render children when routes = [selfName, recordId], array type', () => {
    const mockFetch = jest.fn().mockImplementation(() => Promise.resolve());
    const wrapper = shallow(<WrapperComponent
      type="array"
      routes={['posts', 'post1']}
      rootValue={fromJS([{
        [UNIQUE_ID]: 'post1',
        title: 'Title'
      }])}
      name="posts"
      refId={new RefId("posts")}
      params={{}}
      fetch={mockFetch}
      renderChildren={mockRenderChildren}
    />);
    return wrapper.instance().updateState().then(() => {
      expect(wrapper.instance().state).toMatchObject({
        renderType: 1,
        canRender: true,
        restRoutes: []
      });
    });
  });

  it('render children when routes = [selfName], array type, create operation', () => {
    const mockFetch = jest.fn().mockImplementation(() => Promise.resolve());
    const wrapper = shallow(<WrapperComponent
      type="array"
      routes={['posts']}
      rootValue={fromJS([{
        [UNIQUE_ID]: 'post1',
        title: 'Title'
      }])}
      name="posts"
      refId={new RefId("posts")}
      params={{op: 'create'}}
      fetch={mockFetch}
      renderChildren={mockRenderChildren}
    />);
    return wrapper.instance().updateState().then(() => {
      expect(wrapper.instance().state).toMatchObject({
        renderType: 1,
        canRender: true,
        restRoutes: []
      });
    });
  });
});