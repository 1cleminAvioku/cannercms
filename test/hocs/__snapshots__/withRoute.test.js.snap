// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withRoute should render 1`] = `""`;
